### PR TITLE
HW2

### PR DESCRIPTION
**Pull request by DaniilBoiko**

### CODE REVIEW

File: src/network/mt_blocking/ServerImpl.cpp [hunk 1]
```diff
@@ -91,59 +93,171 @@ void ServerImpl::Join() {
...
+
+            // Configure read timeout
+            {
+                struct timeval tv;
+                tv.tv_sec = _tv_sec;
+                tv.tv_usec = 0;
+                setsockopt(client_socket, SOL_SOCKET, SO_RCVTIMEO, (const char *) &tv, sizeof tv);
+            }
+
+            if (_max_workers == connections.size()) {
```
Comment (blumix) [inline src/network/mt_blocking/ServerImpl.cpp]: тут может быть рейс, нужен lock

File: src/network/mt_blocking/ServerImpl.cpp [hunk 2]
```diff
@@ -79,6 +79,8 @@ void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
 // See Server.h
 void ServerImpl::Stop() {
     running.store(false);
+    assert(_thread.joinable());
+    _thread.join();
```
Comment (blumix) [inline src/network/mt_blocking/ServerImpl.cpp:84]: клиенты могут писать бесконечно
ты сначала ждешь завершения, а потом закрываешь сокет.

File: src/network/mt_blocking/ServerImpl.cpp [hunk 3]
```diff
@@ -91,59 +93,171 @@ void ServerImpl::Join() {
...
+
+        _logger->debug("Waiting threads to finish...");
+
+
+        // Cleanup on exit...
+        std::unique_lock<std::mutex> lock(connection_mutex);
+        cv.wait(lock, [this]{ return (this -> connections.empty()); });
+
+        _logger->warn("Network stopped");
+        close(_server_socket);
```
Comment (blumix) [inline src/network/mt_blocking/ServerImpl.cpp:156]: делаешь close () дважды

**Review by blumix — CHANGES_REQUESTED** [review]

File: src/network/mt_blocking/ServerImpl.cpp [hunk 4]
```diff
@@ -80,70 +80,184 @@ void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
...
+        } else {
+            throw std::runtime_error(std::string(strerror(errno)));
         }
+    } catch (std::runtime_error &ex) {
+        _logger->error("Failed to process connection on descriptor {}: {}", client_socket, ex.what());
     }
 
-    // Cleanup on exit...
-    _logger->warn("Network stopped");
+    close(client_socket);
```
Comment (blumix) [inline src/network/mt_blocking/ServerImpl.cpp]: тут сокет отдается системе, и в другом потоке может быть заменен экземпляр класса std::thread, лежащий в connections. Это приведет к паденю

File: src/network/mt_blocking/ServerImpl.cpp [hunk 5]
```diff
@@ -80,70 +80,184 @@ void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
...
+    command_to_execute.reset();
+    argument_for_command.resize(0);
+    parser.Reset();
+
+    // Remove itself from connections dict
+    std::lock_guard<std::mutex> lock(connection_mutex);
+    connections.erase(client_socket);
```
Comment (blumix) [inline src/network/mt_blocking/ServerImpl.cpp:258]: тут ты удаляешь экземпляр класса, связанный с потоком, а потом что-то продолжаешь в нем делать

**Review by blumix — CHANGES_REQUESTED** [review]

File: src/network/mt_blocking/ServerImpl.cpp [hunk 6]
```diff
@@ -80,70 +80,187 @@ void ServerImpl::Start(uint16_t port, uint32_t n_accept, uint32_t n_workers) {
...
-    // Cleanup on exit...
-    _logger->warn("Network stopped");
+    command_to_execute.reset();
+    argument_for_command.resize(0);
+    parser.Reset();
+
+    // Remove itself from connections dict
+    std::lock_guard<std::mutex> lock(connection_mutex);
+    
+    connections[client_socket].get().detach();
```
Comment (blumix) [inline src/network/mt_blocking/ServerImpl.cpp]: вместо
` connections[client_socket].get().detach();`
можно писать 
` connections[client_socket]->detach();`
